flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java

```diff
@@ -0,0 +1,214 @@
+            throw new SavepointCreationException(
+                    "Unable to stop with savepoint. Error while creating the savepoint.",
+                    throwable);
+        }
+        return completedCheckpoint.getExternalPointer();
+    }
+
+    private String handleStopWithSavepointOperationResult(
+            JobStatusAndSavepointPath statusAndPath, Throwable throwable) {
+        // filter out exception from first stage to show correct root cause
+        if (throwable != null) {
+            if (throwable instanceof SavepointCreationException) {
+                throw (SavepointCreationException) throwable;
+            } else {
+                throw new CompletionException(
+                        "Error while stopping job after creating savepoint", throwable);
+            }
+        }
+
+        if (statusAndPath.getStatus() == JobStatus.FINISHED) {
+            // ensure that all vertices from the execution graph are in state FINISHED. There might
+            // be cases where only some subgraphs of the job are finished.
+            if (!areAllVerticesFinished()) {
+                if (getLogger().isDebugEnabled()) {
+                    getLogger().debug("Printing vertex states:");
+                    for (Execution e : getExecutionGraph().getRegisteredExecutions().values()) {
+                        getLogger().debug("{} = {}", e.getVertex(), e.getState());
+                    }
+                }
+                final FlinkException exception =
+                        new FlinkException(
+                                "Not all vertices of job are finished. Stop with savepoint operation failed.");
+                context.runIfState(this, () -> handleAnyFailure(exception));
+                throw new CompletionException("Unable to stop with savepoint", exception);
+            }
+            context.runIfState(
+                    this,
+                    () ->
+                            context.goToFinished(
+                                    ArchivedExecutionGraph.createFrom(getExecutionGraph())));
+            return statusAndPath.getPath();
+        } else {
+            throw new CompletionException(
+                    new FlinkException(
+                            "Reached state "
+                                    + statusAndPath.getStatus()
+                                    + " instead of FINISHED "));
+        }
+    }
+
+    private boolean areAllVerticesFinished() {
+        return getExecutionGraph().getRegisteredExecutions().values().stream()
+                .noneMatch(e -> e.getState() != ExecutionState.FINISHED);
+    }
+
+    @Override
+    protected void handleAnyFailure(@Nonnull Throwable cause) {
+        Preconditions.checkState(context.isState(this), "Assuming StopWithSavepoint state");
+        // restart the checkpoint coordinator if stopWithSavepoint failed.
+        startCheckpointScheduler(checkpointCoordinator);
```

Comment:
We could say that we do this whenever we leave this state in any other state than `Finished`.